opus_test.go

```diff
@@ -91,6 +91,178 @@ func TestCodecFloat32(t *testing.T) {
 	}
 }
 
+func TestCodecFEC(t *testing.T) {
+	// Create bogus input sound
+	const G4 = 391.995
+	const SAMPLE_RATE = 48000
+	const FRAME_SIZE_MS = 60
+	const FRAME_SIZE = SAMPLE_RATE * FRAME_SIZE_MS / 1000
+	const PACKET_SIZE = 480 // 10ms packet
+	pcm := make([]int16, 0)
+
+	enc, err := NewEncoder(SAMPLE_RATE, 1, AppVoIP)
+	if err != nil || enc == nil {
+		t.Fatalf("Error creating new encoder: %v", err)
+	}
+	enc.SetPacketLossPerc(30)
+	enc.SetInBandFEC(1)
+
+	dec, err := NewDecoder(SAMPLE_RATE, 1)
+	if err != nil || dec == nil {
+		t.Fatalf("Error creating new decoder: %v", err)
+	}
+
+	mono := make([]int16, FRAME_SIZE)
+	addSine(mono, SAMPLE_RATE, G4)
+
+	encodedData := make([][]byte, FRAME_SIZE/PACKET_SIZE)
+	for i, idx := 0, 0; i < len(mono); i += PACKET_SIZE {
+		data := make([]byte, 1000)
+		n, err := enc.Encode(mono[i:i+PACKET_SIZE], data)
+		if err != nil {
+			t.Fatalf("Couldn't encode data: %v", err)
+		}
+		data = data[:n]
+		encodedData[idx] = data
+		idx++
+	}
+
+	lost := false
+	for i := 0; i < len(encodedData); i++ {
+		// "Dropping" packets 2 and 4
+		if i == 2 || i == 4 {
+			lost = true
+			continue
+		}
+		if lost {
+			samples, err := dec.LastPacketDuration()
+			if err != nil {
+				t.Fatalf("Couldn't get last packet duration: %v", err)
+			}
+
+			pcmBuffer := make([]int16, samples)
+			if err = dec.DecodeFEC(encodedData[i], pcmBuffer); err != nil {
+				t.Fatalf("Couldn't decode data: %v", err)
+			}
+			pcm = append(pcm, pcmBuffer...)
+			lost = false
+		}
+
+		pcmBuffer := make([]int16, FRAME_SIZE)
+		n, err := dec.Decode(encodedData[i], pcmBuffer)
+		if err != nil {
+			t.Fatalf("Couldn't decode data: %v", err)
+		}
+		pcmBuffer = pcmBuffer[:n]
+		if len(pcmBuffer) != n {
+			t.Fatalf("Length mismatch: %d samples in, %d out", len(pcmBuffer), n)
```

Comment:
I'm not sure about this error, by the way. the length of the pcmBuffer does not seem to be the amount of samples fed to the encoder in this pass? especially after truncating to :n. What exactly is being tested, here?

Reply:
The idea of the test is to lose two packets (I guess we can do it random but to be honest it was just easier like this).  All of the packets have to be decoded without fec, that's where this line comes from.
In addition to that the packets coming after the "lost ones" have to be decoded with fec on (there you can see that I don't check for the length but just the error).

Reply:
I agree the test logic itself is ok, but this specific guard is a tautology. after `a = a[:n]`, you will always `len(a) ≤ n`, and because you start out with `len(a) = FRAME_SIZE ⋙ n` this is a tautology. You might be trying to test e.g. `if n != PACKET_SIZE`?

Reply:
TBH I just copy/pasted it from the TestCodec test, but I don't mind changing it to `n!=PACKET_SIZE` or actually `n!=FRAME_SIZE` since I changed it.

Reply:
That's fine :) don't forget to change the error message. Currently you're testing for PACKET_SIZE and n, but reporting len(pcmBuffer) and n.